### ISSUE TITLE
[BUG]Model instability and NaN/High MAPE on CUDA (Windows) — Suggest explicit torch.cuda.synchronize() after .to('cpu')

### ISSUE THREAD

**Issue by pinkfloyd798** [unit8co/darts#2828]
**Dear Darts Developers,

I am experiencing significant differences and instability when running rolling forecasts with Darts neural models (especially TCNModel, but also NHiTSModel and TiDEModel) on Windows using PyTorch with CUDA support, compared to CPU-only PyTorch builds.

System configuration:
OS: Windows 10 Pro, 64-bit

GPU: NVIDIA RTX 3060 (12 GB)

Driver: NVIDIA 552.22 (latest WHQL)

CUDA toolkit: CUDA 12.1

Python: 3.12 (also tested with 3.10 and 3.11, Anaconda)

Darts version: 0.35.0

PyTorch versions: 2.5.1+cu121, 2.7.1+cu121 (and CPU builds for reference)

Issue details:
On CPU-only PyTorch, all models work reliably; MAPE is consistent (e.g. 3–4%), rolling forecasts finish without errors, and results match my reference runs.

On GPU/CUDA PyTorch:

TCNModel sometimes produces NaN, extremely high MAPE, or random errors (Appended TimeSeries must start one (time) step after current one).

Other models (NHiTS, TiDE) sometimes return higher error than CPU build on identical data and seeds.

The same data, code, and seeds are used, and all conda environments are 100% isolated.

Switching back to CPU PyTorch resolves all issues.

I am aware that the root cause could be in PyTorch/CUDA itself (and have opened an issue with PyTorch as well), but I suspect that the way Darts handles tensor transfers from GPU to CPU may be a contributing factor.

Possible root cause (forum/community insight):
Community members suggest that tensor.to('cpu') is asynchronous on CUDA builds, and further CPU operations (e.g., converting tensors to numpy or pandas, or any CPU-based API) can occur before the transfer has actually finished, especially on Windows.

The proposed workaround is to call:

python
Copy
Edit
tensor = tensor.to('cpu')
torch.cuda.synchronize()
immediately after transferring to CPU, before any downstream operation (like .numpy(), .tolist(), etc).

This is a known PyTorch pattern and may prevent race conditions or data inconsistency that can manifest as random NaN, errors, or inconsistent metrics.

Request:
Can you please review all code paths where Darts transfers tensors from GPU to CPU, especially in predict(), result serialization, and rolling forecast logic?

If possible, add an explicit torch.cuda.synchronize() after every .to('cpu') operation and before passing tensors to numpy/pandas/other CPU libraries.

At the very least, could you document this as a known issue or provide a workaround for users running Darts on CUDA+Windows?

I am happy to provide further diagnostics, reproducible code, or test experimental fixes if needed.
Thank you for your great work on Darts — it is an amazing library for time series forecasting!

Best regards,
Vladimir Malbašić

**Comment by dennisbader** [comment]
Hi @pinkfloyd798, our `TorchForecastinModel` framework is built on top of PyTorch Lightning (which is built on top of PyTorch). 

The only place where we move the predictions to CPU is in the line below:

https://github.com/unit8co/darts/blob/b71851b3e899db891fdf106b151f006e97b91722/darts/models/forecasting/pl_forecasting_module.py#L396

I pushed branch [fix/tensor_gpu_to_cpu_sync](https://github.com/unit8co/darts/tree/fix/tensor_gpu_to_cpu_sync) which changes this line and let's Lightning handle the device moving.

Can you try installing that branch from a fresh environment and see if it fixes your issue?

E.g. install the branch with below:

```
pip install git+https://github.com/unit8co/darts.git@fix/tensor_gpu_to_cpu_sync
```

If it doesn't fix the issue then I would suggest to ask also on the PyTorch Lightning side (see their GitHub [here](https://github.com/Lightning-AI/pytorch-lightning)).

**Comment by dennisbader** [comment]
Hi @pinkfloyd798, were you able to test it? :)

**Comment by pinkfloyd798** [comment]
Subject: Feedback on Experimental GPU Fix — Stability Improved, Some Differences Remain

Message:

Dear Darts Team,

Thank you for your prompt response and for providing the experimental branch fix/tensor_gpu_to_cpu_sync. I have tested it thoroughly under my previous setup (Windows 10 + RTX 3060 + CUDA 12.1 + PyTorch 2.5.1) and I wanted to share my findings:

 Stability Significantly Improved
After installing the experimental branch, I can confirm that the issues I previously reported no longer occur:

No more NaN values or random spikes in MAPE during rolling forecasts.

The error Appended TimeSeries must start one (time) step after current one has not appeared again.

Grid searches now complete without crashing or producing anomalous results.

This confirms that the fix addressing asynchronous GPU-to-CPU transfers (i.e., using tensor.to('cpu') followed by torch.cuda.synchronize()) is effective — especially on Windows systems where race conditions are more prevalent.

 Performance Still Diverges From CPU
While stability has improved, I noticed that models still produce different MAPE values between the GPU and CPU runs with identical data, seeds, and parameters:

NHiTSModel and TiDEModel consistently show slightly higher MAPE on GPU compared to CPU.

Interestingly, TCNModel performs better on GPU with the same hyperparameters.

This suggests that while instability has been mitigated, some inherent non-determinism in CUDA operations (e.g., due to cuDNN, floating-point precision) may still lead to slight metric divergence.

 Next Steps
I plan to test additional reproducibility strategies (such as enabling torch.backends.cudnn.deterministic = True and setting seeds with seed_everything) in the next round of experiments. I’ll report back with results once I have concrete findings.

I truly appreciate your work on Darts — it's an outstanding library for time series forecasting, and I'm happy to see issues like this being addressed so promptly. Let me know if you'd like me to share reproducible code, diagnostics, or further benchmark results across CPU/GPU platforms.

Kind regards

**Comment by dennisbader** [comment]
Great, thanks a lot for the confirmation and nice words @pinkfloyd798 :)

It is expected that there are some differences between GPU and CPU. As you also saw, it's not that GPU always leads to worse results for the same seed, so all good in my opinion.

In that case I'll merge #2829 and close this issue once done 🚀
